File: scripts/refresh_fixtures.py
# SPDX-License-Identifier: Apache-2.0
"""Public synthetic keys, callbacks and data only."""
import json
from copy import deepcopy
from pathlib import Path
from durable_fixtures import Fixture as DurableFixture, normalize
from psp_cdl_core.crypto import sign_envelope
from psp_cdl_llmproxy import RefreshingLlmLoop, prompt_context
SUITE=json.loads((Path(__file__).resolve().parents[1]/"conformance/vectors/llm/refresh-0.1.json").read_text(encoding="utf-8"))

class Fixture(DurableFixture):
    def __init__(self,settings=None):
        settings=settings or {}
        super().__init__({"continueTurn":True,**settings,"base":{**settings.get("base",{}),"promptRefresh":not settings.get("disabledRefresh")}})
        self.refresh_requests,self.refresh_events=[],[]
        self.template={"version":"1.0.0","text":"System one.","timestamp":900,"expires":1200,"attributes":{"refresh-policy":"interval|expiration","refresh-interval":"2","refresh-grace":"100"},**settings.get("initial",{})}
        self.base.flags["onInvoke"]=lambda:self.base.flags.update(now=self.flags["afterToolTime"]) if self.flags.get("afterToolTime") else None
        commit=self.base.backend.commit
        def commit_with_ack(*args):
            ok=commit(*args)
            if ok and self.flags.get("failAfterPromptCommit") and any(w["body"].get("profile")=="PSP-PROMPT-REFRESH-0.1" for w in args[2]):raise RuntimeError("PRIVATE_ACK_DETAIL")
            return ok
        self.base.backend.commit=commit_with_ack
        try:self.loop=RefreshingLlmLoop(self.base.store,self.base.gate,self,self.provider,{"postCompletion":"lockdown"})
        except Exception:
            self.close()
            raise
    def sign(self,b,t):return sign_envelope(t["text"],{"algorithm":"hmac-sha256","signatureVersion":"2.0","secretId":"test-signing-key","timestamp":t["timestamp"],"expires":t["expires"],"version":t["version"],"sectionType":"system","contentType":"text","trustLevel":t.get("trustLevel",2),"attributes":{**prompt_context(b),**t["attributes"]}},self.key)
    def prompt(self,p,b):return self.sign(b,{**self.template,**self.flags.get("bindingTemplate",{})})
    def refresh(self,p,b,request):
        self.refresh_requests.append(deepcopy(request))
        if self.flags.get("onRefresh"):self.flags["onRefresh"](p,b,request)
        if self.flags.get("throwRefresh"):raise RuntimeError("PRIVATE_REFRESH_DETAIL")
        for flag,key in (("refreshCancel","cancelled"),("refreshDrift","drift"),("refreshRevoke","revoked")):
            if self.flags.get(flag):self.base.flags[key]=True
        if self.flags.get("refreshRace"):
            try:self.base.update()
            except Exception as exc:
                if getattr(exc,"code",None)!="STATE_BUSY":raise
                self.refresh_events.append({"signal":"STATE_BUSY"})
        self.template={**self.template,"version":self.flags.get("refreshVersion","1.0.1"),"text":self.flags.get("refreshText","System refreshed."),"timestamp":self.flags.get("refreshTimestamp",self.base.flags["now"]),"expires":self.flags.get("refreshExpires",self.base.flags["now"]+500),**self.flags.get("replacement",{})}
        e=self.sign(b,self.template)
        if self.flags.get("tamperedRefresh"):e["data"]="tampered"
        if self.flags.get("wrongRefreshScope"):e["signature"]["attributes"]["session-id"]="other"
        return e
    def authorize_refresh(self,p,b,previous,candidate):return not self.flags.get("denyInitial" if b["trigger"]=="initial" else "denyRefresh")
    def audit_refresh(self,p,event):
        self.refresh_events.append(event)
        if self.flags.get("throwRefreshAudit"):raise RuntimeError("PRIVATE_AUDIT_DETAIL")
        return not self.flags.get("denyRefreshAudit")
    def invoke(self,*args):
        result=super().invoke(*args)
        if self.flags.get("afterProviderTime"):self.base.flags["now"]=self.flags["afterProviderTime"]
        return result

def run_case(case):
    f=None;steps=[]
    try:
        f=Fixture(case.get("settings"))
        for step in case.get("steps",[{"action":"run"}]):
            f.flags.update(step.get("flags",{}));f.base.flags.update(step.get("baseFlags",{}))
            options={**f.options,**step.get("options",{})}
            if step["action"]=="restart":f.loop=RefreshingLlmLoop(f.base.store,f.base.gate,f,f.provider,{"postCompletion":"lockdown"})
            try:
                value=f.loop.recover(step.get("token","test-owner"),f.base.session["sessionId"],step.get("requestId","turn-1"),options) if step["action"]=="recover" else f.loop.run(step.get("token","test-owner"),f.base.session["sessionId"],step.get("request",{"message":"Read synthetic data."}),options)
                steps.append({"code":"OK","value":value})
            except Exception as exc:steps.append({"code":getattr(exc,"code","UNEXPECTED_ERROR")})
        state=f.base.store.execute(f.base.actor,{"action":"getSession","sessionId":f.base.session["sessionId"]})
        prompt=None
        try:prompt=f.base.store.execute(f.base.actor,{"action":"getPromptState","sessionId":f.base.session["sessionId"]})
        except Exception as exc:
            if getattr(exc,"code",None)!="NOT_FOUND":raise
        return normalize({"steps":steps,"codes":[s["code"] for s in steps],"released":sum(s["code"]=="OK" for s in steps),**f.stats(),"sessionVersion":state["version"],"status":state["status"],"prompt":prompt,"turnCount":prompt["state"]["turnCount"] if prompt else None,"refreshCount":prompt["state"]["refreshCount"] if prompt else None,"version":prompt["state"]["version"] if prompt else None,"refreshCalls":len(f.refresh_requests),"refreshRequests":f.refresh_requests,"events":f.refresh_events,"requests":f.requests,"commands":f.commands},f.base.session["sessionId"])
    except Exception as exc:return {"code":getattr(exc,"code","UNEXPECTED_ERROR")}
    finally:
        if f:f.close()

File: scripts/refresh-fixtures.mjs
// SPDX-License-Identifier: Apache-2.0
// Public synthetic keys, callbacks and data only.
import {readFileSync} from 'node:fs';
import {fixture as durableFixture,normalize} from './durable-fixtures.mjs';
import {signEnvelope} from '@psp-cdl/core/crypto';
import {RefreshingLlmLoop,promptContext} from '@psp-cdl/llmproxy';
export const suite=JSON.parse(readFileSync(new URL('../conformance/vectors/llm/refresh-0.1.json',import.meta.url),'utf8'));
export async function fixture(settings={}) {
  const f=await durableFixture({continueTurn:true,...settings,base:{...settings.base,promptRefresh:!settings.disabledRefresh}}),refreshRequests=[],refreshEvents=[];
  let template={version:'1.0.0',text:'System one.',timestamp:900,expires:1200,attributes:{'refresh-policy':'interval|expiration','refresh-interval':'2','refresh-grace':'100'},...settings.initial};
  const sign=(b,t)=>signEnvelope(t.text,{algorithm:'hmac-sha256',signatureVersion:'2.0',secretId:'test-signing-key',timestamp:t.timestamp,expires:t.expires,version:t.version,sectionType:'system',contentType:'text',trustLevel:t.trustLevel??2,attributes:{...promptContext(b),...t.attributes}},new Uint8Array(32).fill(19));
  const host={...f.host,
    prompt:(_p,b)=>sign(b,{...template,...f.flags.bindingTemplate}),
    refresh:async(p,b,request)=>{
      refreshRequests.push(structuredClone(request));
      if(f.flags.onRefresh)await f.flags.onRefresh(p,b,request);
      if(f.flags.throwRefresh)throw Error('PRIVATE_REFRESH_DETAIL');
      if(f.flags.refreshCancel)f.baseFlags.cancelled=true;
      if(f.flags.refreshDrift)f.baseFlags.drift=true;
      if(f.flags.refreshRevoke)f.baseFlags.revoked=true;
      if(f.flags.refreshRace){try{await f.update();}catch(e){if(e.code!=='STATE_BUSY')throw e;refreshEvents.push({signal:'STATE_BUSY'});}}
      template={...template,version:f.flags.refreshVersion??'1.0.1',text:f.flags.refreshText??'System refreshed.',timestamp:f.flags.refreshTimestamp??f.baseFlags.now,expires:f.flags.refreshExpires??f.baseFlags.now+500,...f.flags.replacement};
      const e=sign(b,template);
      if(f.flags.tamperedRefresh)e.data='tampered';
      if(f.flags.wrongRefreshScope)e.signature.attributes['session-id']='other';
      return e;
    },
    authorizeRefresh:(_p,b,_previous,_candidate)=>!(b.trigger==='initial'?f.flags.denyInitial:f.flags.denyRefresh),
    auditRefresh:(_p,event)=>{refreshEvents.push(event);if(f.flags.throwRefreshAudit)throw Error('PRIVATE_AUDIT_DETAIL');return !f.flags.denyRefreshAudit;}
  };
  f.baseFlags.onInvoke=()=>{if(f.flags.afterToolTime)f.baseFlags.now=f.flags.afterToolTime;};
  const commit=f.backend.commit.bind(f.backend);
  f.backend.commit=async(...args)=>{
    const ok=await commit(...args);
    if(ok&&f.flags.failAfterPromptCommit&&args[2].some(w=>w.body.profile==='PSP-PROMPT-REFRESH-0.1'))throw Error('PRIVATE_ACK_DETAIL');
    return ok;
  };
  const invoke=f.provider.invoke;
  const provider={...f.provider,invoke:async(...args)=>{const result=await invoke(...args);if(f.flags.afterProviderTime)f.baseFlags.now=f.flags.afterProviderTime;return result;}};
  try{return {...f,host,provider,refreshRequests,refreshEvents,loop:new RefreshingLlmLoop(f.store,f.gate,host,provider,{postCompletion:'lockdown'})};}
  catch(e){f.close();throw e;}
}
export async function runCase(c){
  let f;const steps=[];
  try{
    f=await fixture(c.settings);
    for(const step of c.steps??[{action:'run'}]){
      Object.assign(f.flags,step.flags??{});Object.assign(f.baseFlags,step.baseFlags??{});
      const options={...f.options,...step.options};
      if(step.action==='restart')f.loop=new RefreshingLlmLoop(f.store,f.gate,f.host,f.provider,{postCompletion:'lockdown'});
      try{const value=step.action==='recover'?await f.loop.recover(step.token??'test-owner',f.session.sessionId,step.requestId??'turn-1',options):await f.loop.run(step.token??'test-owner',f.session.sessionId,step.request??{message:'Read synthetic data.'},options);steps.push({code:'OK',value});}
      catch(e){steps.push({code:e.code??'UNEXPECTED_ERROR'});}
    }
    const state=await f.store.execute(f.actor,{action:'getSession',sessionId:f.session.sessionId});
    let prompt=null;try{prompt=await f.store.execute(f.actor,{action:'getPromptState',sessionId:f.session.sessionId});}catch(e){if(e.code!=='NOT_FOUND')throw e;}
    return normalize({steps,codes:steps.map(s=>s.code),released:steps.filter(s=>s.code==='OK').length,...f.stats(),sessionVersion:state.version,status:state.status,prompt,turnCount:prompt?.state.turnCount??null,refreshCount:prompt?.state.refreshCount??null,version:prompt?.state.version??null,refreshCalls:f.refreshRequests.length,refreshRequests:f.refreshRequests,events:f.refreshEvents,requests:f.requests,commands:f.commands},f.session.sessionId);
  }catch(e){return {code:e.code??'UNEXPECTED_ERROR'};}finally{f?.close();}
}
